### ISSUE TITLE
Breakage with recent nightly

### ISSUE THREAD

**Issue by svmnotn** [serde-rs/aster#79]
```
[ aster]$ rustup show
active toolchain
----------------
nightly-x86_64-unknown-linux-gnu (default)
rustc 1.10.0-nightly (7bddce693 2016-05-27)

[ aster]$ cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling aster v0.16.0 (file:///home/svmnotn/data/dev/rand/aster)
src/expr.rs:1387:65: 1387:75 error: mismatched types [E0308]
src/expr.rs:1387         self.builder.build_expr_kind(ast::ExprKind::Loop(block, self.label))
                                                                                 ^~~~~~~~~~
src/expr.rs:1387:65: 1387:75 help: run `rustc --explain E0308` to see a detailed explanation
src/expr.rs:1387:65: 1387:75 note: expected type `std::option::Option<syntax::codemap::Spanned<syntax::ast::Ident>>`
src/expr.rs:1387:65: 1387:75 note:    found type `std::option::Option<syntax::ast::Ident>`
src/method.rs:43:45: 43:66 error: no associated item named `Static` found for type `syntax::ast::SelfKind` in the current scope
src/method.rs:43             explicit_self: respan(DUMMY_SP, ast::SelfKind::Static),
                                                             ^~~~~~~~~~~~~~~~~~~~~
src/method.rs:90:13: 90:34 error: no associated item named `Static` found for type `syntax::ast::SelfKind` in the current scope
src/method.rs:90             ast::SelfKind::Static => { }
                             ^~~~~~~~~~~~~~~~~~~~~
src/method.rs:92:15: 92:46 error: this pattern has 3 fields, but the corresponding variant has 2 fields [E0023]
src/method.rs:92             | ast::SelfKind::Region(_, _, id)
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/method.rs:92:15: 92:46 help: run `rustc --explain E0023` to see a detailed explanation
src/method.rs:95:31: 95:49 error: no associated item named `new_self` found for type `syntax::ast::Arg` in the current scope
src/method.rs:95                     let arg = ast::Arg::new_self(self.span, self.self_mutable, id);
                                               ^~~~~~~~~~~~~~~~~~
src/method.rs:108:13: 108:26 error: structure `syntax::ast::MethodSig` has no field named `explicit_self`
src/method.rs:108             explicit_self: self.explicit_self,
                              ^~~~~~~~~~~~~
src/method.rs:192:30: 192:51 error: no associated item named `Static` found for type `syntax::ast::SelfKind` in the current scope
src/method.rs:192         self.build_self_kind(ast::SelfKind::Static)
                                               ^~~~~~~~~~~~~~~~~~~~~
src/method.rs:196:30: 196:51 error: no associated item named `Static` found for type `syntax::ast::SelfKind` in the current scope
src/method.rs:196         self.build_self_kind(ast::SelfKind::Static)
                                               ^~~~~~~~~~~~~~~~~~~~~
src/method.rs:200:51: 200:68 error: mismatched types [E0308]
src/method.rs:200         self.build_self_kind(ast::SelfKind::Value("self".to_ident()))
                                                                    ^~~~~~~~~~~~~~~~~
src/method.rs:200:51: 200:68 help: run `rustc --explain E0308` to see a detailed explanation
src/method.rs:200:51: 200:68 note: expected type `syntax::ast::Mutability`
src/method.rs:200:51: 200:68 note:    found type `syntax::ast::Ident`
src/method.rs:204:30: 208:10 error: this function takes 2 parameters but 3 parameters were supplied [E0061]
src/method.rs:204         self.build_self_kind(ast::SelfKind::Region(
                                               ^
src/method.rs:204:30: 208:10 help: run `rustc --explain E0061` to see a detailed explanation
src/method.rs:214:30: 218:10 error: this function takes 2 parameters but 3 parameters were supplied [E0061]
src/method.rs:214         self.build_self_kind(ast::SelfKind::Region(
                                               ^
src/method.rs:214:30: 218:10 help: run `rustc --explain E0061` to see a detailed explanation
src/method.rs:222:30: 226:10 error: this function takes 2 parameters but 3 parameters were supplied [E0061]
src/method.rs:222         self.build_self_kind(ast::SelfKind::Region(
                                               ^
src/method.rs:222:30: 226:10 help: run `rustc --explain E0061` to see a detailed explanation
src/method.rs:232:30: 236:10 error: this function takes 2 parameters but 3 parameters were supplied [E0061]
src/method.rs:232         self.build_self_kind(ast::SelfKind::Region(
                                               ^
src/method.rs:232:30: 236:10 help: run `rustc --explain E0061` to see a detailed explanation
src/method.rs:250:58: 250:75 error: mismatched types [E0308]
src/method.rs:250         self.build_self_kind(ast::SelfKind::Explicit(ty, "self".to_ident()))
                                                                           ^~~~~~~~~~~~~~~~~
src/method.rs:250:58: 250:75 help: run `rustc --explain E0308` to see a detailed explanation
src/method.rs:250:58: 250:75 note: expected type `syntax::ast::Mutability`
src/method.rs:250:58: 250:75 note:    found type `syntax::ast::Ident`
src/pat.rs:248:37: 248:90 error: this function takes 3 parameters but 2 parameters were supplied [E0061]
src/pat.rs:248         self.builder.build_pat_kind(ast::PatKind::TupleStruct(self.path, Some(self.pats)))
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/pat.rs:248:37: 248:90 help: run `rustc --explain E0061` to see a detailed explanation
src/pat.rs:423:37: 423:54 error: no associated item named `Tup` found for type `syntax::ast::PatKind` in the current scope
src/pat.rs:423         self.builder.build_pat_kind(ast::PatKind::Tup(self.pats))
```

**Comment by hartmut** [comment]
with the nightly build from 27.05.2016 it works:
nightly-2016-05-27-x86_64-unknown-linux-gnu

**Comment by svmnotn** [comment]
That is weird because that was the nightly that rustup showed I was using.

Ok, when I run `rustup toolchain install nightly-2016-05-27`, rustup installs the rustc from the day before 2016-05-26, not the one from the 27th. The rustc from the 26th builds aster just fine.

**Comment by svmnotn** [comment]
The same issue exists in `rustc 1.11.0-nightly (7746a334d 2016-05-28)`

**Comment by erickt** [comment]
Hello all,

I'm not quite done testing it, but I've (mostly) fixed it in #80. There were a lot of changes, so I want to first check it against the rest of the ecosystem before pushing it out.

**Comment by tikue** [comment]
Thanks for your hard work @erickt !
